[GNA] Fix exception handling on wait and infer

marcinkusm opened the pull request:
### Details:
- added exception handling for wait to allow to start new inference if error happened on wait
- added closing unfinished queued requests if error happen on wait or enqueue

### Tickets:
 -  98800

nkogteva reviewed (commented):
In general, I'm against the having one more wrapper around infer request. Instead of it it is better to reuse common plugin approach with task executor, but the priority of this task is also questionble for me @dorloff  It would be nice to have a fix for particular issue without global refactoring and additional GNA specific wrappers which complificate the code.

marcinkusm commented:
@nkogteva I agree that task executor would be better, but I am not sure what would be a potential effort.
In my PR didn't add any new wrapper. I have just moved code from header to source file, due the fact there was not need to keep it in header.

It took me a while to understand how does GNAInferRequest class behave this is why I have split particular functionalities to separate methods to easy understanding. 

I  have also added missing cleanup of enqueued inferences for in case of issue for one of sub request in the situation when model would be splitted, otherwise this request would be handing in gna lib until destruction of plugin instance.

nkogteva commented:
> 



> @nkogteva I agree that task executor would be better, but I am not sure what would be a potential effort. In my PR didn't add any new wrapper. I have just moved code from header to source file, due the fact there was not need to keep it in header.
> 
> It took me a while to understand how does GNAInferRequest class behave this is why I have split particular functionalities to separate methods to easy understanding.
> 
> I have also added missing cleanup of enqueued inferences for in case of issue for one of sub request in the situation when model would be splitted, otherwise this request would be handing in gna lib until destruction of plugin instance.

Ok, I see. But difference is huge enough and it is hard to realize what is actually was added and what is just moved from header. Can we split it?

marcinkusm commented:
> > 
> 
> > @nkogteva I agree that task executor would be better, but I am not sure what would be a potential effort. In my PR didn't add any new wrapper. I have just moved code from header to source file, due the fact there was not need to keep it in header.
> > It took me a while to understand how does GNAInferRequest class behave this is why I have split particular functionalities to separate methods to easy understanding.
> > I have also added missing cleanup of enqueued inferences for in case of issue for one of sub request in the situation when model would be splitted, otherwise this request would be handing in gna lib until destruction of plugin instance.
> 
> Ok, I see. But difference is huge enough and it is hard to realize what is actually was added and what is just moved from header. Can we split it?

I did it in one commit due the fact in first case I have fixed issue with subrequests and than realized that changes in GNAInferRequest  are also needed, but you are right it should be slit into separate changes.
Is it fine for you to split these changes into two commits, one with containing only move code to source file and second with rest of the changes?

kbruniec commented on src/plugins/intel_gna/src/request/worker_impl.cpp:
```diff
@@ -42,23 +43,36 @@ Gna2Model* WorkerImpl::model() {
 
-    for (auto& subrequest : modelSubrequests_) {
-        subrequest->enqueue();
+    try {
+        for (auto& subrequest : modelSubrequests_) {
+            subrequest->enqueue();
+        }
+    } catch (const std::exception& e) {
+        ov::intel_gna::log::error() << "Exception when enqueueing inference: " << e.what() << std::endl;
+        cleanup_unfinished_submodules();
+        throw;
```
Do we have test for that?

marcinkusm replied:
Reworked worker. 
Tests for it are done unit tests gna_request_worker_impl_test.cpp

kbruniec commented on src/plugins/intel_gna/src/request/worker_impl.cpp:
```diff
@@ -42,23 +43,36 @@ Gna2Model* WorkerImpl::model() {
-        }
+    try {
+        // iterate over all configurations for requst
+        for (auto& subrequest : modelSubrequests_) {
+            if (!subrequest->isPending()) {
+                continue;
+            }
 
-        if (subrequest->wait(timeoutMilliseconds) == RequestStatus::kPending) {
-            pending = true;
+            if (subrequest->wait(timeoutMilliseconds) == RequestStatus::kPending) {
+                pending = true;
+            }
         }
+    } catch (const std::exception& e) {
+        ov::intel_gna::log::error() << "Exception when executiong wait: " << e.what() << std::endl;
+        pending = false;
+        cleanup_unfinished_submodules();
+        throw;
     }
```
1. i would get rid of try/catch and introduce RequestStatus::kCompletedWithError
2. cleanup_unfinished_submodules() would be not needed
3. We dont need to throw from this function, we can return kCompletedWithError and throw from outside function

marcinkusm replied:
1. Ok
2. We still  need try to cleanup due the fact we could receive kPending.
3. Ok

kbruniec replied:
Ad2. 
make sense,
...
however if we get booth: pending & kCompletedWithError (for subrequests) we could just return pending (overal status) and do the de-facto cleanup at the next call to wait and throw kCompletedWithError when all pending subrequests are completed,

this scenario is not likely to happen (combines three unusual situations big model + pending +error from wait) so making the code simpler is IMO the most important

marcinkusm replied:
I have changed worker to not throw for enqueue and wait, but that required some changes in gna_plugin.cpp

kbruniec commented on src/plugins/intel_gna/src/request/worker_impl.cpp:
```diff
@@ -113,5 +127,18 @@ void WorkerImpl::check_if_free() {
     }
 }
 
+void WorkerImpl::cleanup_unfinished_submodules() {
```
IMO we should get rid of this function,
otherwise the function name could be:
cleanup_model_subrequests

marcinkusm replied:
I have modified it.

kbruniec reviewed (commented):
In general looks fine,
still IMO it could be simplified by some changes proposed...

nkogteva commented on src/plugins/intel_gna/src/gna_infer_request.cpp:
```diff
@@ -0,0 +1,174 @@
+    // if result is ok we set kRequestIndexCompleted to not execute request if it is not
+    // in the queue.
+    auto result_request_index = result ? kRequestIndexCompleted : kRequestIndexInvalid;
+    SetRequestIndex(result_request_index);
```
Maybe it is better to put it inside `infer_call`? Because in case of exception you set request id

marcinkusm replied:
I am not sure if get your comment properly, because I think in case of exception request id will be set to invalid and exception will be rethrown.
Due the fact infer_call is executed in context of 
```
void GNAInferRequest::CallCleanupAndRethrowOnException(std::function<void()>&& function_to_invoke) {
    try {
        function_to_invoke();
    } catch (...) {
        // need to preserve invalid state here to avoid next Wait() from clearing it
        // and next rethrow issue.
        SetRequestIndex(kRequestIndexInvalid);
        throw;
    }
}
```

nkogteva replied:
> I am not sure if get your comment properly, because I think in case of exception request id will be set to invalid and exception will be rethrown. Due the fact `infer_call` is executed in context of
> 
> ```
> void GNAInferRequest::CallCleanupAndRethrowOnException(std::function<void()>&& function_to_invoke) {
>     try {
>         function_to_invoke();
>     } catch (...) {
>         // need to preserve invalid state here to avoid next Wait() from clearing it
>         // and next rethrow issue.
>         SetRequestIndex(kRequestIndexInvalid);
>         throw;
>     }
> }
> ```

Yes, but there is no sense to keep it outside `infer_call`.

marcinkusm replied:
You comment is right, but on the other hand I wanted to make this lambdas as simple as possible and consistent across infer/start_async. Intention was to do there only calls to `plg`.

nkogteva replied:
> You comment is right, but on the other hand I wanted to make this lambdas as simple as possible and consistent across infer/start_async. Intention was to do there only calls to `plg`.

for `queue_call` it is not true, look unaligned

nkogteva commented on src/plugins/intel_gna/src/request/worker_impl.cpp:
```diff
@@ -39,28 +40,41 @@ Gna2Model* WorkerImpl::model() {
-    check_if_free();
+bool WorkerImpl::enqueueRequest() {
+    if (!isFree()) {
+        ov::intel_gna::log::warning() << "Trying to propagte on busy request with id: " << representingIndex_;
```
```suggestion
        ov::intel_gna::log::warning() << "Trying to propagate on busy request with id: " << representingIndex_;
```

marcinkusm commented on src/plugins/intel_gna/src/gna_infer_request.cpp:
```diff
@@ -68,33 +78,75 @@ void GNAInferRequest::StartAsyncImpl() {
 
-    if (waitStatus == RequestStatus::kPending) {
+    if (millis_timeout < 0) {
+        IE_THROW(ParameterMismatch);
```
Something like?
`"Invalid timeout value in milisecons: " << millis_timeout << "!".`

nkogteva replied:
> Something like? `"Invalid timeout value in milliseconds: " << millis_timeout << "!".`

yep

marcinkusm commented on src/plugins/intel_gna/src/gna_infer_request.cpp:
```diff
@@ -68,33 +78,75 @@ void GNAInferRequest::StartAsyncImpl() {
     }
 
+    if (IsRequestCompleted()) {
+        return InferenceEngine::OK;
```
If I get your comment properly, you mean that before checking if request is completed I should check if timeout is valid?

nkogteva replied:
> If I get your comment properly, you mean that before checking if request is completed I should check if timeout is valid?

yes

marcinkusm commented:
Comments fixed

nkogteva commented on src/plugins/intel_gna/tests/unit/gna_infer_request_test.cpp:
```diff
@@ -0,0 +1,204 @@
+        return function;
+    }
+
+    void SetExpectsForLoadNetworhAndShutDown(std::vector<std::vector<uint8_t>>& data) {
```
```suggestion
    void SetExpectsForLoadNetworkAndShutDown(std::vector<std::vector<uint8_t>>& data) {
```

kbruniec reviewed (approved):
LGTM

kbruniec reviewed (commented):
WOuld be good to execute
updateGnaPerfCounters & per_request_diagnostics
even when Gna2RequestWait return error

marcinkusm commented:
> Would be good to execute:
> 
> * `updateGnaPerfCounters()`
> * `per_request_diagnostics` block
>   in `waitForRequest()` even when `Gna2RequestWait()` returns error and `waitForRequest()` throws from `checkGna2Status()`
Done
